src/Parser.jl

```diff
@@ -48,267 +123,277 @@ function _count_before{T<:AbstractString}(haystack::T, needle::Char, _end::Int)
     return count
 end
 
-# Prints an error message with an indicator to the source
-function _error(message::AbstractString, ps::ParserState)
-    lines = _count_before(ps.str, '\n', ps.s)
+
+# Throws an error message with an indicator to the source
+function _error(message::AbstractString, ps::MemoryParserState)
+    orig = Compat.UTF8String(ps.utf8data)
+    lines = _count_before(orig, '\n', ps.s)
     # Replace all special multi-line/multi-space characters with a space.
-    strnl = replace(ps.str, r"[\b\f\n\r\t\s]", " ")
+    strnl = replace(orig, r"[\b\f\n\r\t\s]", " ")
     li = (ps.s > 20) ? ps.s - 9 : 1 # Left index
-    ri = min(ps.e, ps.s + 20)       # Right index
+    ri = min(endof(orig), ps.s + 20)       # Right index
     error(message *
       "\nLine: " * string(lines) *
       "\nAround: ..." * strnl[li:ri] * "..." *
       "\n           " * (" " ^ (ps.s - li)) * "^\n"
     )
 end
 
+function _error(message::AbstractString, ps::StreamingParserState)
+    error("$message\n ...when parsing byte with value '$(current(ps))'")
+end
+
 # PARSING
 
-function parse_array{T<:AbstractString}(ps::ParserState{T}, dictT::Type)
-    incr(ps) # Skip over the '['
-    _array = Any[]
-    chomp_space(ps)
-    charat(ps)==']' && (incr(ps); return _array) # Check for empty array
-    while true # Extract values from array
-        v = parse_value(ps, dictT) # Extract value
-        push!(_array, v)
-        # Eat up trailing whitespace
-        chomp_space(ps)
-        c = charat(ps)
-        if c == ','
-            incr(ps)
-            continue
-        elseif c == ']'
-            incr(ps)
-            break
-        else
-            _error("Unexpected char: " * string(c), ps)
-        end
+"""
+Given a `ParserState`, after possibly any amount of whitespace, return the next
+parseable value.
+"""
+function parse_value(ps::ParserState, dictT::Type)
+    chomp_space!(ps)
+
+    @inbounds byte = byteat(ps)
+    if byte == STRING_DELIM
+        parse_string(ps)
+    elseif isjsondigit(byte) || byte == MINUS_SIGN
+        parse_number(ps)
+    elseif byte == OBJECT_BEGIN
+        parse_object(ps, dictT)
+    elseif byte == ARRAY_BEGIN
+        parse_array(ps, dictT)
+    else
+        parse_jsconstant(ps::ParserState)
     end
-    return _array
 end
 
-function parse_object{T<:AbstractString}(ps::ParserState{T}, dictT::Type)
-    # Handle case where type is not parameterized yet with key type
-    keyT = dictT.parameters[1]
-    if typeof(keyT) == TypeVar
-        parse_object(ps, dictT, AbstractString, dictT{AbstractString, Any}())
+function parse_jsconstant(ps::ParserState)
+    c = advance!(ps)
+    if c == LATIN_T
+        skip!(ps, LATIN_R, LATIN_U, LATIN_E)
+        true
+    elseif c == LATIN_F
+        skip!(ps, LATIN_A, LATIN_L, LATIN_S, LATIN_E)
+        false
+    elseif c == LATIN_N
+        skip!(ps, LATIN_U, LATIN_L, LATIN_L)
+        nothing
```

Comment:
I think it would be good to add a comment to each condition specifying what it's testing for, e.g.

```julia
    if c == LATIN_T  #  "true"
        ...
```
It's decodable right now... but a comment can just be read.